README.md
# sprout-detection

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Project structure](#project-structure)
- [Setup](#setup)
  - [Installation](#installation)
  - [Required Python Packages](#required-python-packages)
- [Usage](#usage)
  - [color_picker](#color_picker)
  - [hsv_filter](#hsv_filter)
  - [main](#main)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

Project of signal, image and video @ University of Trento for the detection of sprouts.


## Project structure

    sprout-detection
    ├── (images)            [ignored folder where you can add your images and then change the "image.extension" in the files with the name of your images]
    └── src
        ├── image_utils               [package that contains the utils for images]
        |    └── standard_image_operations               [library which contains the main operations on images]
        ├── color_picker              [script to pick HSV colors and define ranges]
        ├── hsv_filter                [script to try our standard HSV filter on different kind of images]
        └── main                      [script to extract the relevant edges of the image using canny and sobel, also using an HSV filtering approach]


## Setup

### Installation

This repository can be cloned using the command:

```bash
    git clone https://github.com/samuelbortolin/sprout-detection.git
```


### Required Python Packages

Required Python packages can be installed using the command:

```bash
    pip install -r requirements.txt
```


## Usage

### color_picker

Change the path with the image to analyze and run the script:
* in the first phase it allows you to pick pixels from an image and then clicking `q` it returns the range of HSV values that contains all the selected pixels;
* in the second phase it allows you to build an HSV filter using trackbars to modify the value of hue, saturation and value, clicking a button it updates HSV filtered image and clicking `q` it returns the range of HSV values selected with the trackbars.

The script can be run using one of the two set of commands:

```bash
    cd src
    python3 color_picker.py
```

or

```bash
    cd src
    python3 -m color_picker
```

### hsv_filter

Change the path with the different types of images containing flowers, leaves or branches to analyze and run the script: it performs an isolation of the color for these images using our standard HSV filter.

The script can be run using one of the two set of commands:

```bash
    cd src
    python3 hsv_filter.py
```

or

```bash
    cd src
    python3 -m hsv_filter
```

### main

Change the path with the image to analyze and run the script: it extracts the relevant edges of the image using canny and sobel. It also tries an HSV color filtering approach and extracts the relevant edges of the HSV filtered image.

The script can be run using one of the two set of commands:

```bash
    cd src
    python3 main.py
```

or

```bash
    cd src
    python3 -m main
```


CHANGELOG.md
# change-log

## 1.2.3

* Renamed imshow window names


## 1.2.2

* Updated documentation and minor code refactoring


## 1.2.1

* Code refactoring


## 1.2.0

* Code refactoring and updated image_operations library
* Recomputed the threshold for hsv edges


## 1.1.1

* Minor fixes and refactoring of the main
* Refactored the hsv color request


## 1.1.0

* Refactor of the main
* Updated documentation


## 1.0.0

* Released a first version of the project for detecting the spouts, trying to use edge detectors, hsv filters and other approaches
* Built an image_utils package for standard image operations


src/main.py
from __future__ import absolute_import, annotations

from copy import deepcopy

import cv2 as cv
import numpy as np

from image_utils.standard_image_operations import StandardImageOperations as SIO


image_path = "../images/image.extension"


if __name__ == "__main__":

    image = cv.imread(image_path)
    image = SIO.rescale_image(image)
    cv.imshow("original image", image)
    greyscale_image = cv.cvtColor(image, cv.COLOR_BGR2GRAY)

    # try an edge detector approach
    sobel_edges = SIO.sobel_edges(greyscale_image)
    sobel_edges_after_percentile_removal = SIO.remove_percentile(sobel_edges)
    canny_edges = SIO.canny_edges(greyscale_image, SIO.find_canny_best_threshold(greyscale_image))

    canny_bitwise_and_sobel_after_percentile_removal = np.zeros((len(image), len(image[0])))
    for i, item_i in enumerate(canny_edges):
        for j, item_j in enumerate(item_i):
            if item_j.any() > 0 and sobel_edges_after_percentile_removal[i, j] > 0:
                canny_bitwise_and_sobel_after_percentile_removal[i, j] = int(255)
            else:
                canny_bitwise_and_sobel_after_percentile_removal[i, j] = int(0)
    canny_bitwise_and_sobel_after_percentile_removal = canny_bitwise_and_sobel_after_percentile_removal.astype(np.uint8)
    cv.imshow("Sobel-Canny edges approach", canny_bitwise_and_sobel_after_percentile_removal)

    image_and_edges = deepcopy(image)
    for i, item_i in enumerate(canny_bitwise_and_sobel_after_percentile_removal):
        for j, item_j in enumerate(item_i):
            if item_j != 0:
                image_and_edges[i][j] = (0, 0, 255)
    cv.imshow("Sobel-Canny edges on original image", image_and_edges)

    # try to use area
    contour_tuple = cv.findContours(canny_bitwise_and_sobel_after_percentile_removal, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
    contours = SIO.grab_contours(contour_tuple)
    contour_image = deepcopy(canny_bitwise_and_sobel_after_percentile_removal)

    for contour in contours:
        cv.drawContours(contour_image, [contour], 0, (100, 5, 10), 10)

    image_with_contours_area = deepcopy(image)
    for i, item_i in enumerate(contour_image):
        for j, item_j in enumerate(item_i):
            if item_j.all() == 0:
                image_with_contours_area[i][j] = (0, 0, 0)
    cv.imshow("contours area approach", image_with_contours_area)

    # try hsv color filter after edges
    color = input("What do you want to analyze? Type f for flowers, l for leaves or b for branches: ")
    hsv_filtered_image = SIO.get_hsv_mask(image, cv.cvtColor(image, cv.COLOR_BGR2HSV), color)
    cv.imshow("hsv-filtered image", hsv_filtered_image)

    hsv_edges = SIO.canny_edges(cv.cvtColor(hsv_filtered_image, cv.COLOR_BGR2GRAY), SIO.find_canny_best_threshold(cv.cvtColor(hsv_filtered_image, cv.COLOR_BGR2GRAY)))
    hsv_bitwise_and_canny_and_sobel = np.zeros((len(image), len(image[0])))
    for i, item_i in enumerate(cv.bitwise_and(canny_edges, hsv_edges)):
        for j, item_j in enumerate(item_i):
            if item_j.any() > 0 and sobel_edges_after_percentile_removal[i, j] > 0:
                hsv_bitwise_and_canny_and_sobel[i, j] = int(255)
            else:
                hsv_bitwise_and_canny_and_sobel[i, j] = int(0)
    hsv_bitwise_and_canny_and_sobel = hsv_bitwise_and_canny_and_sobel.astype(np.uint8)
    cv.imshow("hsv-filtering edges + Sobel-Canny edges approach", hsv_bitwise_and_canny_and_sobel)

    image_and_edges_hsv = deepcopy(image)
    hsv_filtered_image_and_edges_hsv = deepcopy(hsv_filtered_image)
    for i, item_i in enumerate(hsv_bitwise_and_canny_and_sobel):
        for j, item_j in enumerate(item_i):
            if item_j.all() != 0:
                image_and_edges_hsv[i][j] = (0, 0, 255)
                hsv_filtered_image_and_edges_hsv[i][j] = (0, 0, 255)
    cv.imshow("hsv-filtering edges + Sobel-Canny edges approach on original image", image_and_edges_hsv)
    cv.imshow("hsv-filtering edges + Sobel-Canny edges approach on the hsv-filtered image", hsv_filtered_image_and_edges_hsv)

    cv.waitKey()
    cv.destroyAllWindows()


src/hsv_filter.py
from __future__ import absolute_import, annotations

import cv2 as cv

from image_utils.standard_image_operations import StandardImageOperations as SIO


flowers_image_path = "../images/flowers_image.extension"
leaves_image_path = "../images/leaves_image.extension"
branches_image_path = "../images/branches_image.extension"


if __name__ == "__main__":

    # hsv color filter applied to flowers
    image = cv.imread(flowers_image_path)
    image = SIO.rescale_image(image)
    hsv_image = cv.cvtColor(image, cv.COLOR_BGR2HSV)
    cv.imshow("result after applying hsv mask to image with flowers", SIO.get_hsv_mask(image, hsv_image, "f"))

    # hsv color filter applied to leaves
    image = cv.imread(leaves_image_path)
    image = SIO.rescale_image(image)
    hsv_image = cv.cvtColor(image, cv.COLOR_BGR2HSV)
    cv.imshow("result after applying hsv mask to image with leaves", SIO.get_hsv_mask(image, hsv_image, "l"))

    # hsv color filter applied to branches
    image = cv.imread(branches_image_path)
    image = SIO.rescale_image(image)
    hsv_image = cv.cvtColor(image, cv.COLOR_BGR2HSV)
    cv.imshow("result after applying hsv mask to image with branches", SIO.get_hsv_mask(image, hsv_image, "b"))

    cv.waitKey()
    cv.destroyAllWindows()


src/sobel_canny.py
from __future__ import absolute_import, annotations

from copy import deepcopy

import cv2 as cv
import numpy as np

from image_utils.image_opertions import StandardImageOperations as SIO


image_path = "../images/image.extension"

percentage = 85
scale = 1
delta = 0
depth = cv.CV_16S
n_derivative = 2


if __name__ == "__main__":

    image = cv.imread(image_path)
    image = SIO.rescale_image(image)
    color = input("What do you want to analyze? Write w for white/flowers, g for green/leaves or b for brown/branches: ")
    hsv_filtered_image = SIO.get_hsv_mask(image, cv.cvtColor(image, cv.COLOR_BGR2HSV), color)
    greyscale_image = cv.cvtColor(image, cv.COLOR_BGR2GRAY)
    blurred_image = cv.GaussianBlur(image, (3, 3), 0)
    greyscale_blurred_image = cv.cvtColor(blurred_image, cv.COLOR_BGR2GRAY)

    # try with sobel
    gradient_x = cv.Sobel(greyscale_blurred_image, depth, n_derivative, 0, ksize=3, scale=scale, delta=delta, borderType=cv.BORDER_DEFAULT)
    gradient_y = cv.Sobel(greyscale_blurred_image, depth, 0, n_derivative, ksize=3, scale=scale, delta=delta, borderType=cv.BORDER_DEFAULT)
    abs_gradient_x = cv.convertScaleAbs(gradient_x)
    abs_gradient_y = cv.convertScaleAbs(gradient_y)
    gradient_image = cv.addWeighted(abs_gradient_x, 0.5, abs_gradient_y, 0.5, 0)
    cv.imshow("sobel gradient image", gradient_image)
    kernel = np.ones((3, 3), np.uint8)
    close_after_sobel = cv.morphologyEx(gradient_image, cv.MORPH_CLOSE, kernel)
    cv.imshow("close after sobel", close_after_sobel)
    open_after_close_on_sobel = cv.morphologyEx(close_after_sobel, cv.MORPH_OPEN, kernel)
    cv.imshow("open after close on sobel", open_after_close_on_sobel)

    # try to remove a percentile
    positive_values_array = []
    for i, item_i in enumerate(open_after_close_on_sobel):
        for j, item_j in enumerate(item_i):
            if item_j > 0:
                positive_values_array.append(item_j)
    percentile = np.percentile(positive_values_array, percentage)

    for i, item_i in enumerate(open_after_close_on_sobel):
        for j, item_j in enumerate(item_i):
            if item_j <= percentile:
                open_after_close_on_sobel[i, j] = 0
    cv.imshow("percentile removed", open_after_close_on_sobel)

    # try a canny edge detector approach
    previous_threshold = SIO.canny_elements(greyscale_image, 25)
    previous_delta = SIO.canny_elements(greyscale_image, 24) - previous_threshold
    picked_threshold = 0
    for i in range(26, 1000):
        new_threshold = SIO.canny_elements(greyscale_image, i)
        delta = previous_threshold - new_threshold
        previous_threshold = new_threshold
        if (delta + previous_delta) < 50:
            picked_threshold = i
            break
        previous_delta = delta

    # show the original and the resulting image
    cv.imshow("original image", image)
    edges_on_image = SIO.canny_on_image(greyscale_image, picked_threshold, image)
    cv.imshow("relevant edges of the image", edges_on_image)

    edged_image = SIO.canny_edges(greyscale_image, picked_threshold)
    bitwise_and_percentile_canny = np.zeros((len(image), len(image[0])))
    for i, item_i in enumerate(edged_image):
        for j, item_j in enumerate(item_i):
            if item_j.any() > 0 and open_after_close_on_sobel[i, j] > 0:
                bitwise_and_percentile_canny[i, j] = int(255)
            else:
                bitwise_and_percentile_canny[i, j] = int(0)
    bitwise_and_percentile_canny = bitwise_and_percentile_canny.astype(np.uint8)
    cv.imshow("bitwise_and percentile canny", bitwise_and_percentile_canny)

    # try to use also hsv
    bitwise_and_on_image_hsv = deepcopy(hsv_filtered_image)
    for i, item_i in enumerate(bitwise_and_percentile_canny):
        for j, item_j in enumerate(item_i):
            if item_j.all() != 0:
                bitwise_and_on_image_hsv[i][j] = (0, 0, 255)
    cv.imshow("bitwise_and percentile canny on image hsv", bitwise_and_on_image_hsv)

    # try to use area
    contour_tuple = cv.findContours(bitwise_and_percentile_canny, cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
    contours = SIO.grab_contours(contour_tuple)
    contour_image = deepcopy(bitwise_and_percentile_canny)

    for contour in contours:
        cv.drawContours(contour_image, [contour], 0, (100, 5, 10), 10)

    image_with_contours_area = deepcopy(image)
    for i, item_i in enumerate(contour_image):
        for j, item_j in enumerate(item_i):
            if item_j.all() == 0:
                image_with_contours_area[i][j] = (0, 0, 0)
    cv.imshow("image with contours area", image_with_contours_area)

    cv.waitKey()
    cv.destroyAllWindows()


src/color_picker.py
from __future__ import absolute_import, annotations

import cv2 as cv
import numpy as np

from image_utils.standard_image_operations import StandardImageOperations as SIO


image_path = "../images/image.extension"

colors = []

H_low = 0
H_high = 179
S_low = 0
S_high = 255
V_low = 0
V_high = 255


def on_mouse_click(event, x, y, flags, hsv_image):
    # mouse click function to store the hsv value

    if event == cv.EVENT_LBUTTONUP:
        colors.append(hsv_image[y, x].tolist())


def callback(x):
    # trackbar callback function assigning trackbar position value to H, S, V high and low variables

    global H_low, H_high, S_low, S_high, V_low, V_high
    H_low = cv.getTrackbarPos("low_H", "control_H")
    H_high = cv.getTrackbarPos("high_H", "control_H")
    S_low = cv.getTrackbarPos("low_S", "control_S")
    S_high = cv.getTrackbarPos("high_S", "control_S")
    V_low = cv.getTrackbarPos("low_V", "control_V")
    V_high = cv.getTrackbarPos("high_V", "control_V")


if __name__ == "__main__":

    # selects points to build a mask
    while True:
        image = cv.imread(image_path)
        image = SIO.rescale_image(image)
        hsv_image = cv.cvtColor(image, cv.COLOR_BGR2HSV)
        if colors:
            cv.putText(hsv_image, str(colors[-1]), (10, 50), cv.FONT_HERSHEY_PLAIN, 2, (0, 0, 0), 2)
        cv.imshow("hsv image", hsv_image)
        cv.setMouseCallback("hsv image", on_mouse_click, hsv_image)

        if cv.waitKey(100) & 0xFF == ord("q"):
            break

    cv.destroyAllWindows()

    min_h = min(c[0] for c in colors)
    min_s = min(c[1] for c in colors)
    min_v = min(c[2] for c in colors)
    max_h = max(c[0] for c in colors)
    max_s = max(c[1] for c in colors)
    max_v = max(c[2] for c in colors)

    lower_bound = np.array([min_h, min_s, min_v])
    upper_bound = np.array([max_h, max_s, max_v])
    print(lower_bound, upper_bound)

    mask = cv.inRange(hsv_image, lower_bound, upper_bound)
    result_image = cv.bitwise_and(image, image, mask=mask)
    cv.imshow("mask", mask)
    cv.imshow("result image", result_image)

    # show complementary on h
    lower_bound = np.array([0, min_s, min_v])
    upper_bound = np.array([min_h, max_s, max_v])
    mask_lower = cv.inRange(hsv_image, lower_bound, upper_bound)
    result_lower_image = cv.bitwise_and(image, image, mask=mask_lower)

    lower_bound = np.array([max_h, min_s, min_v])
    upper_bound = np.array([179, max_s, max_v])
    mask_upper = cv.inRange(hsv_image, lower_bound, upper_bound)
    result_upper_image = cv.bitwise_and(image, image, mask=mask_upper)

    complementary_h_mask = np.bitwise_or(mask_lower, mask_upper)
    result_complementary_h_image = cv.bitwise_and(image, image, mask=complementary_h_mask)
    cv.imshow("complementary H mask", complementary_h_mask)
    cv.imshow("result complementary H image", result_complementary_h_image)

    cv.waitKey()
    cv.destroyAllWindows()

    # create separate windows and trackbars for high,low H,S,V
    cv.namedWindow("control_H", 2)
    cv.createTrackbar("low_H", "control_H", 0, 179, callback)
    cv.createTrackbar("high_H", "control_H", 179, 179, callback)
    cv.resizeWindow("control_H", 400, 20)

    cv.namedWindow("control_S", 2)
    cv.createTrackbar("low_S", "control_S", 0, 255, callback)
    cv.createTrackbar("high_S", "control_S", 255, 255, callback)
    cv.resizeWindow("control_S", 400, 20)

    cv.namedWindow("control_V", 2)
    cv.createTrackbar("low_V", "control_V", 0, 255, callback)
    cv.createTrackbar("high_V", "control_V", 255, 255, callback)
    cv.resizeWindow("control_V", 400, 20)

    # move trackbars to build a mask
    while True:
        image = cv.imread(image_path)
        image = SIO.rescale_image(image)
        hsv_image = cv.cvtColor(image, cv.COLOR_BGR2HSV)

        lower_bound = np.array([H_low, S_low, V_low])
        upper_bound = np.array([H_high, S_high, V_high])

        mask = cv.inRange(hsv_image, lower_bound, upper_bound)
        result_image = cv.bitwise_and(image, image, mask=mask)
        cv.imshow("hsv mask", mask)
        cv.imshow("hsv filtered image", result_image)

        if cv.waitKey() & 0xFF == ord("q"):
            break

    cv.destroyAllWindows()
    print(lower_bound, upper_bound)


src/play_with_HSV.py
from __future__ import absolute_import, annotations

import cv2 as cv
import numpy as np


image_flowers_path = "../media/ROBI_VT01_20200507_3.png"
image_leaves_path = "../media/FOTO1ROBI_MI02_20200422.jpg"
image_branches_path = "../media/image.jpg"


def apply_mask(original_image: np.ndarray, hsv_image: np.ndarray, lower_bound: np.ndarray, upper_bound: np.ndarray) -> np.ndarray:
    # get the image after applying hsv mask based on a range of hsv colors

    mask = cv.inRange(hsv_image, lower_bound, upper_bound)
    # mask = cv.morphologyEx(mask, cv.MORPH_OPEN, (5, 5), iterations=1)
    return cv.bitwise_and(original_image, original_image, mask=mask)


def get_hsv_mask(original_image: np.ndarray, hsv_image: np.ndarray, color: str) -> np.ndarray:
    # get the image after applying hsv mask based on color of interest

    if color == "w":
        lower_bound = np.array([0, 0, 75])
        upper_bound = np.array([90, 255, 255])
        lower_image = apply_mask(original_image, hsv_image, lower_bound, upper_bound)
        lower_bound = np.array([115, 0, 75])
        upper_bound = np.array([179, 255, 255])
        upper_image = apply_mask(original_image, hsv_image, lower_bound, upper_bound)
        return cv.bitwise_or(lower_image, upper_image)
    elif color == "g":  # values to be estimated using the color_picker and then tested with play_with_HSV
        lower_bound = np.array([20, 0, 0])
        upper_bound = np.array([80, 255, 255])
    elif color == "b":  # values to be estimated using the color_picker and then tested with play_with_HSV
        lower_bound = np.array([0, 0, 0])
        upper_bound = np.array([30, 255, 255])
    else:
        lower_bound = np.array([0, 0, 0])
        upper_bound = np.array([180, 255, 255])
    return apply_mask(original_image, hsv_image, lower_bound, upper_bound)


def rescale_image(image_to_rescale: np.ndarray, target_number_of_pixels: int = 100000) -> np.ndarray:
    # rescale image all to the same standard number of pixels

    scale = (len(image_to_rescale[0]) * len(image_to_rescale) / target_number_of_pixels) ** (1 / 2)
    return cv.resize(image_to_rescale, (0, 0), fx=(1 / scale), fy=(1 / scale))


if __name__ == "__main__":

    # x flowers
    image = cv.imread(image_flowers_path)
    image = rescale_image(image)
    hsv_image = cv.cvtColor(image, cv.COLOR_BGR2HSV)
    cv.imshow("result after applying hsv mask", get_hsv_mask(image, hsv_image, "w"))
    cv.waitKey()
    cv.destroyAllWindows()

    # x leaves
    image = cv.imread(image_leaves_path)
    image = rescale_image(image)
    hsv_image = cv.cvtColor(image, cv.COLOR_BGR2HSV)
    cv.imshow("result after applying hsv mask", get_hsv_mask(image, hsv_image, "g"))
    cv.waitKey()
    cv.destroyAllWindows()

    # x branches
    image = cv.imread(image_branches_path)
    image = rescale_image(image)
    hsv_image = cv.cvtColor(image, cv.COLOR_BGR2HSV)
    cv.imshow("result after applying hsv mask", get_hsv_mask(image, hsv_image, "b"))
    cv.waitKey()
    cv.destroyAllWindows()


src/image_utils/image_opertions.py
from __future__ import absolute_import, annotations

from copy import deepcopy
from typing import List

import cv2 as cv
import numpy as np


class StandardImageOperations:

    ratio = 3  # we can try to set a high threshold instead of using this ratio
    kernel_size = 3  # we can try to understand what is that
    scale = 1
    delta = 0
    depth = cv.CV_16S
    n_derivative = 2
    percentage = 85

    @staticmethod
    def rescale_image(image_to_rescale: np.ndarray, target_number_of_pixels: int = 100000) -> np.ndarray:
        """
        Rescale the image to a target number of pixels
        """

        scale = (len(image_to_rescale[0]) * len(image_to_rescale) / target_number_of_pixels) ** (1 / 2)
        return cv.resize(image_to_rescale, (0, 0), fx=(1 / scale), fy=(1 / scale))

    @staticmethod
    def grab_contours(contour_tuple: tuple) -> List[np.ndarray]:
        """
        Grab a list contours from the tuple resulting from cv.findContours
        """

        # for OpenCV v2.4, v4-official
        if len(contour_tuple) == 2:
            return contour_tuple[0]
        # for OpenCV v3
        elif len(contour_tuple) == 3:
            return contour_tuple[1]

    @staticmethod
    def find_canny_best_threshold(greyscale_image: np.ndarray):
        """
        Find the best threshold for canny edge detector
        """

        canny_on_previous_threshold = StandardImageOperations.canny_elements(greyscale_image, 25)
        canny_previous_delta = StandardImageOperations.canny_elements(greyscale_image, 24) - canny_on_previous_threshold
        picked_threshold = 0
        for threshold_value in range(26, 1000):
            canny_on_new_threshold = StandardImageOperations.canny_elements(greyscale_image, threshold_value)
            canny_new_delta = canny_on_previous_threshold - canny_on_new_threshold
            canny_on_previous_threshold = canny_on_new_threshold
            if (canny_new_delta + canny_previous_delta) < 50:
                picked_threshold = threshold_value
                break
            canny_previous_delta = canny_new_delta
        return picked_threshold

    @staticmethod
    def canny_edges(greyscale_image: np.ndarray, low_threshold: int) -> np.ndarray:
        """
        Apply canny edge detector and return the edges
        """

        greyscale_blurred_image = cv.GaussianBlur(greyscale_image, (3, 3), 0)
        return cv.Canny(greyscale_blurred_image, low_threshold, low_threshold * StandardImageOperations.ratio, StandardImageOperations.kernel_size)

    @staticmethod
    def canny_elements(greyscale_image: np.ndarray, low_threshold: int) -> int:
        """
        Apply canny edge detector and return the sum of elements in the mask generated by the edges
        """

        greyscale_blurred_image = cv.GaussianBlur(greyscale_image, (3, 3), 0)
        detected_edges = cv.Canny(greyscale_blurred_image, low_threshold, low_threshold * StandardImageOperations.ratio, StandardImageOperations.kernel_size)
        mask = detected_edges != 0
        return np.ma.sum(mask)

    @staticmethod
    def canny_on_image(greyscale_image: np.ndarray, low_threshold: int, original_image: np.ndarray) -> np.ndarray:
        """
        Apply canny edge detector and return the edges on the original image
        """

        greyscale_blurred_image = cv.GaussianBlur(greyscale_image, (3, 3), 0)
        detected_edges = cv.Canny(greyscale_blurred_image, low_threshold, low_threshold * StandardImageOperations.ratio, StandardImageOperations.kernel_size)
        mask = detected_edges != 0
        return original_image * (mask[:, :, None].astype(original_image.dtype))

    @staticmethod
    def sobel_edges(greyscale_image: np.ndarray) -> np.ndarray:
        """
        Apply sobel edge detector and return the edges
        """

        greyscale_blurred_image = cv.GaussianBlur(greyscale_image, (3, 3), 0)
        gradient_x = cv.Sobel(greyscale_blurred_image, StandardImageOperations.depth, StandardImageOperations.n_derivative, 0, ksize=StandardImageOperations.kernel_size,
                              scale=StandardImageOperations.scale, delta=StandardImageOperations.delta, borderType=cv.BORDER_DEFAULT)
        gradient_y = cv.Sobel(greyscale_blurred_image, StandardImageOperations.depth, 0, StandardImageOperations.n_derivative, ksize=StandardImageOperations.kernel_size,
                              scale=StandardImageOperations.scale, delta=StandardImageOperations.delta, borderType=cv.BORDER_DEFAULT)
        abs_gradient_x = cv.convertScaleAbs(gradient_x)
        abs_gradient_y = cv.convertScaleAbs(gradient_y)
        gradient_image = cv.addWeighted(abs_gradient_x, 0.5, abs_gradient_y, 0.5, 0)
        kernel = np.ones((3, 3), np.uint8)
        close_after_sobel = cv.morphologyEx(gradient_image, cv.MORPH_CLOSE, kernel)
        return cv.morphologyEx(close_after_sobel, cv.MORPH_OPEN, kernel)

    @staticmethod
    def remove_percentile(greyscale_image: np.ndarray) -> np.ndarray:
        """
        Remove percentile from a grey scale image
        """

        positive_values_array = []
        for i, item_i in enumerate(greyscale_image):
            for j, item_j in enumerate(item_i):
                if item_j > 0:
                    positive_values_array.append(item_j)
        percentile = np.percentile(positive_values_array, StandardImageOperations.percentage)

        image_without_percentile = deepcopy(greyscale_image)
        for i, item_i in enumerate(greyscale_image):
            for j, item_j in enumerate(item_i):
                if item_j <= percentile:
                    image_without_percentile[i, j] = 0

        return image_without_percentile

    @staticmethod
    def apply_hsv_mask(original_image: np.ndarray, hsv_image: np.ndarray, lower_bound: np.ndarray, upper_bound: np.ndarray) -> np.ndarray:
        """
        Apply an hsv mask based on a range of hsv colors on the original image
        """

        mask = cv.inRange(hsv_image, lower_bound, upper_bound)
        # mask = cv.morphologyEx(mask, cv.MORPH_OPEN, (5, 5), iterations=1)
        return cv.bitwise_and(original_image, original_image, mask=mask)

    @staticmethod
    def get_hsv_mask(original_image: np.ndarray, hsv_image: np.ndarray, color: str) -> np.ndarray:
        """
        Applying an hsv mask based on a color of interest on the original image
        """

        if color == "f":
            lower_bound = np.array([0, 0, 50])
            upper_bound = np.array([35, 255, 255])
            lower_image = StandardImageOperations.apply_hsv_mask(original_image, hsv_image, lower_bound, upper_bound)
            lower_bound = np.array([75, 0, 50])
            upper_bound = np.array([90, 255, 255])
            upper_image = StandardImageOperations.apply_hsv_mask(original_image, hsv_image, lower_bound, upper_bound)
            lower_image = cv.bitwise_or(lower_image, upper_image)
            lower_bound = np.array([110, 0, 50])
            upper_bound = np.array([179, 255, 255])
            upper_image = StandardImageOperations.apply_hsv_mask(original_image, hsv_image, lower_bound, upper_bound)
            return cv.bitwise_or(lower_image, upper_image)
        elif color == "l":
            lower_bound = np.array([20, 0, 0])
            upper_bound = np.array([80, 255, 255])
        elif color == "b":
            lower_bound = np.array([0, 0, 0])
            upper_bound = np.array([30, 255, 255])
        else:
            lower_bound = np.array([0, 0, 0])
            upper_bound = np.array([180, 255, 255])
        return StandardImageOperations.apply_hsv_mask(original_image, hsv_image, lower_bound, upper_bound)


src/canny.py
from __future__ import absolute_import, annotations

import cv2 as cv
import numpy as np
from copy import deepcopy
from typing import List

from play_with_HSV import rescale_image, get_hsv_mask


image_path = "../media/20200402_104028_001.jpg"  # in the future we can set the path as argument or env var

ratio = 3  # we can try to set a high threshold instead of using this ratio
kernel_size = 3  # we can try to understand what is that

percentage = 85
scale = 1
delta = 0
depth = cv.CV_16S
n_derivative = 2


def grab_contours(contour_tuple: tuple) -> List[np.ndarray]:
    # in OpenCV v2.4, v4-official
    if len(contour_tuple) == 2:
        return contour_tuple[0]
    # in OpenCV v3
    elif len(contour_tuple) == 3:
        return contour_tuple[1]

def canny_threshold(greyscale_image: np.ndarray, low_threshold: int) -> np.ndarray:
    # apply canny and return the edges

    blurred_image = cv.blur(greyscale_image, (3, 3))
    return cv.Canny(blurred_image, low_threshold, low_threshold * ratio, kernel_size)


def canny_threshold_mask(greyscale_image: np.ndarray, low_threshold: int) -> int:
    # apply canny and return the sum of elements in the mask

    blurred_image = cv.blur(greyscale_image, (3, 3))
    detected_edges = cv.Canny(blurred_image, low_threshold, low_threshold * ratio, kernel_size)
    mask = detected_edges != 0
    return np.ma.sum(mask)


def canny_threshold_with_image(greyscale_image: np.ndarray, low_threshold: int, original_image: np.ndarray) -> np.ndarray:
    # apply canny and return the edges on the image

    blurred_image = cv.blur(greyscale_image, (3, 3))
    detected_edges = cv.Canny(blurred_image, low_threshold, low_threshold * ratio, kernel_size)
    mask = detected_edges != 0
    return original_image * (mask[:, :, None].astype(original_image.dtype))


if __name__ == "__main__":

    image = cv.imread(image_path)
    image = rescale_image(image)
    image_hsv = get_hsv_mask(image, cv.cvtColor(image, cv.COLOR_BGR2HSV), "g")
    greyscale_image = cv.cvtColor(image, cv.COLOR_BGR2GRAY)

    blurred_image = cv.GaussianBlur(image, (3, 3), 0)
    greyscale_blurred_image = cv.cvtColor(blurred_image, cv.COLOR_BGR2GRAY)

    gradient_x = cv.Sobel(greyscale_blurred_image, depth, n_derivative, 0, ksize=3, scale=scale, delta=delta, borderType=cv.BORDER_DEFAULT)
    gradient_y = cv.Sobel(greyscale_blurred_image, depth, 0, n_derivative, ksize=3, scale=scale, delta=delta, borderType=cv.BORDER_DEFAULT)
    abs_gradient_x = cv.convertScaleAbs(gradient_x)
    abs_gradient_y = cv.convertScaleAbs(gradient_y)
    gradient_image = cv.addWeighted(abs_gradient_x, 0.5, abs_gradient_y, 0.5, 0)
    cv.imshow("sobel gradient image", gradient_image)
    kernel = np.ones((3, 3), np.uint8)
    closing = cv.morphologyEx(gradient_image, cv.MORPH_CLOSE, kernel)
    cv.imshow("closing after sobel", closing)
    open_after_closing = cv.morphologyEx(closing, cv.MORPH_OPEN, kernel)
    cv.imshow("open after closing", open_after_closing)

    positiveValuesArray = []
    for i, itemi in enumerate(open_after_closing):
        for j, itemj in enumerate(itemi):
            if itemj > 0:
                positiveValuesArray.append(itemj)
    percentile = np.percentile(positiveValuesArray, percentage)

    for i, itemi in enumerate(open_after_closing):
        for j, itemj in enumerate(itemi):
            if itemj <= percentile:
                open_after_closing[i, j] = 0
    cv.imshow("percentile removed", open_after_closing)
    # try a edge detector approach
    previous_threshold = canny_threshold_mask(greyscale_image, 25)
    previous_delta = canny_threshold_mask(greyscale_image, 24) - previous_threshold
    picked_threshold = 0
    for i in range(26, 1000):
        new_threshold = canny_threshold_mask(greyscale_image, i)
        delta = previous_threshold - new_threshold
        previous_threshold = new_threshold
        if (delta + previous_delta) < 50:
            picked_threshold = i
            break
        previous_delta = delta

    # show the original and the resulting image
    cv.imshow("original image", image)
    edges_on_image = canny_threshold_with_image(greyscale_image, picked_threshold, image)
    cv.imshow("relevant edges of the image", edges_on_image)
    greyscale_image = cv.cvtColor(image, cv.COLOR_BGR2GRAY)
    edged = canny_threshold(greyscale_image, picked_threshold)
    bitwise_and = np.zeros((len(image), len(image[0])))
    for i, itemi in enumerate(edges_on_image):
        for j, itemj in enumerate(itemi):
            if itemj.any() > 0 and open_after_closing[i, j] > 0:
                bitwise_and[i, j] = int(255)
            else:
                bitwise_and[i, j] = int(0)
    bitwise_and = bitwise_and.astype(np.uint8)
    cv.imshow("bitwise", bitwise_and)
    bitwise_and_on_image = deepcopy(image_hsv)
    for i, item_i in enumerate(bitwise_and):
        for j, item_j in enumerate(item_i):
            if item_j.all() != 0:
                bitwise_and_on_image[i][j] = (0, 0, 255)
    cv.imshow("bitwise_and_on_image", bitwise_and_on_image)

    contour_tuple = cv.findContours(bitwise_and.copy(), cv.RETR_EXTERNAL, cv.CHAIN_APPROX_SIMPLE)
    contours = grab_contours(contour_tuple)
    contour_image = bitwise_and.copy()
    area = 0

    for c in contours:
        area += cv.contourArea(c)
        cv.drawContours(contour_image, [c], 0, (100, 5, 10), 10)

    image_from_contours = deepcopy(image)
    for i, item_i in enumerate(contour_image):
        for j, item_j in enumerate(item_i):
            if item_j.all() == 0:
                image_from_contours[i][j] = (0, 0, 0)
    cv.imshow("image_from_contours", image_from_contours)

    cv.imwrite("../test_images/bitwise_edges_on_image6.jpg", bitwise_and_on_image)
    cv.imwrite("../test_images/original_image6.jpg", image)
    cv.imwrite("../test_images/bitwise_canny_percentile6.jpg", bitwise_and)
    cv.imwrite("../test_images/hsv_image6.jpg", image_hsv)
    cv.imwrite("../test_images/bitwise_canny_percentile_area6.jpg", image_from_contours)

    cv.waitKey()
    cv.destroyAllWindows()
